vowpalwabbit/csv_parser/src/parse_example_csv.cc

```diff
@@ -0,0 +1,280 @@
+  std::string ns = all->data_filename;
+  size_t found = ns.find_last_of('.');
+  if (found != std::string::npos) { ns = ns.substr(0, found); }
+
+  unsigned char _index = static_cast<unsigned char>(ns[0]);
+  bool new_index = false;
+
+  _channel_hash = all->example_parser->hasher(ns.data(), ns.length(), all->hash_seed);
+
+  if (ae->feature_space[_index].size() == 0) { new_index = true; }
+
+  ae->feature_space[_index].start_ns_extent(_channel_hash);
+  parse_features(all, ae->feature_space[_index], csv_line, (all->audit || all->hash_inv) ? ns.c_str() : nullptr);
+  if (new_index && ae->feature_space[_index].size() > 0) { ae->indices.push_back(_index); }
+
+  ae->feature_space[_index].end_ns_extent();
+}
+
+void parser::parse_features(VW::workspace* all, features& fs, std::vector<VW::string_view> csv_line, const char* ns)
+{
+  size_t _anon = 0;
+  for (size_t i = 0; i < csv_line.size(); i++)
+  {
+    // Skip label
+    if (label_index == i) { continue; }
+    // Ignore namespace info value
+    float _cur_channel_v = 1.f;
+    uint64_t word_hash;
+    float _v;
+    // Use header line for feature names
+    VW::string_view feature_name = _header[i];
+    VW::string_view string_feature_value = csv_line[i];
+    bool is_feature_float = check_if_float(string_feature_value);
+
+    float float_feature_value = 0.f;
+
+    if (is_feature_float)
+    {
+      float_feature_value = std::stof(string_feature_value.data());
+      _v = _cur_channel_v * float_feature_value;
+    }
+    else
+    {
+      _v = 1;
+    }
+
+    // Case where feature value is string
+    if (!is_feature_float)
+    {
+      // chain hash is hash(feature_value, hash(feature_name, namespace_hash)) & parse_mask
+      word_hash = (all->example_parser->hasher(string_feature_value.data(), string_feature_value.length(),
+                       all->example_parser->hasher(feature_name.data(), feature_name.length(), _channel_hash)) &
+          all->parse_mask);
+    }
+    // Case where feature value is float and feature name is not empty
+    else if (!feature_name.empty())
+    {
+      word_hash =
+          (all->example_parser->hasher(feature_name.data(), feature_name.length(), _channel_hash) & all->parse_mask);
+    }
+    // Case where feature value is float and feature name is empty
+    else
+    {
+      word_hash = _channel_hash + _anon++;
+    }
+
+    // don't add 0 valued features to list of features
+    if (_v == 0) { continue; }
+    fs.push_back(_v, word_hash);
+
+    if ((all->audit || all->hash_inv) && ns != nullptr)
+    {
+      if (!is_feature_float)
+      {
+        fs.space_names.push_back(VW::audit_strings(ns, std::string{feature_name}, std::string{string_feature_value}));
+      }
+      else
+      {
+        fs.space_names.push_back(VW::audit_strings(ns, std::string{feature_name}));
+      }
+    }
+  }
+}
+
+std::vector<VW::string_view> parser::split(VW::string_view sv, std::string ch)
```

Comment:
Quotes allows you to explicitly specify that the value is a string, not a float.

e.g. the BostonHousing.csv example, the value of `chas` are quotes with `"0"` and `"1"` representing each categories / true or false. It will be wrong, if no quotes exists and then the 0 and 1 would be considered as numerical.

Actually I'm not supporting escapes here, the escapes symbols are just a CPP grammar features to represent `"` inside `""` . And so far I haven't dive deep into the quotes pairing issue, if there exist quotes in the suffix or prefix, I will consider it as string definitely. Also I just blindly separate `,` here, if the separated values number doesn't match the features number, an error will be thrown.

Reply:
The reason I ask is that when quotes are used, you need escapes to represent quotes inside of the input. For example: `"this is a quote\"."`, but with escapes you can represent anything.

It does make things hard when datasets contain quotes. The issue with them though is you need to parse them to actually determine when the row ends.

Reply:
Just like I've mentioned before, the CSV separator has a higher priority than anything, including the quotes. As users has plenty of valid characters to choose as the separator, there's no need to escape.

e.g. `1,"this, is.",2`, by `,` separator, we will directly separate the line into four parts. `1`, `"this`, ` is."`, `2`, as the second and third parts violates the quotes pairing, error will be thrown. And to avoid such situation, the separator can be changed into tabs or something else.

For `1,"this is a quote".",2`, after the quotes pairing and removing processing, we still get `1`,`this is a quote".`,`2`.

Well for `1,"this is a quote."",2`, we do need to introduce the escape. Will work on it for escaping `"` and `'` when they are just next to the end of the quotes.

I know it's something that is counter-intuitive, but it will be the easiest way to handle this.

Reply:
Hmm, that is counter intuitive. I guess we just need to be clear - quotes are not treated specially and are just a character.

Reply:
> I guess we just need to be clear - quotes are not treated specially and are just a character.

Yes, you are right. But quotes are sometimes special:
e.g. we need to remove the outer quotes to make multi-classes labels clean, as for Iris2.csv, the label value strings are quoted, if the user just specify `Setosa,Versicolor,Virginica`, it would be wrong (without removing outer quotes), and only `"Setosa","Versicolor","Virginica"` would work, which is not users' general practice. 

And also, as we treat quotes just as a character now, outer quotes pairing are usually a good error indicator. Again, e.g. `1,"this, is.",2`, by `,` separator, we will directly separate the line into four parts. `1`, `"this`, ` is."`, `2`, as the second and third parts violates the outer quotes pairing, error will be thrown. This helps user get their idea that there's a separator conflict since we no longer consider quotes special.

Reply:
Both requiring outer quotes to be paired while also not respecting quoting separators seems extra confusing to me.

So we are sort of understanding quotes but then deciding to explicitly not respect them?

Reply:
Well, that's reasonable. Maybe we can just stop throwing the quotes pairing error to avoid being confusing to users.

For removing the outer quotes to make multi-classes labels clean, when the outer quotes pair, we just remove them, if they don't, we just keep them without throwing any errors. And with that solution, we would benefit the user while not confusing them.

Reply:
I still think that is potentially confusing them. We are changing the meaning of their input in some special case.

If we don't wish to handle quotes at first then I believe we should simply treat them the same as any other character.

It is not unreasonable to require some data preprocessing - especially for a first version of this parser

Reply:
Fine. Then I will comment out codes related to it first, for easier revert if needed later, when if we decide to handle quotes again.

Reply:
Just add an option `--csv_remove_quotes` for auto removing outer quotes if needed, and it would be disabled by default. I guess this would be the best solution so far.